docs/proposals/20191017-control-plane.md

```diff
@@ -0,0 +1,494 @@
+
+  // Ready denotes that the ControlPlane API Server is ready to
+  // recieve requests.
+  Ready bool `json:"ready,omitempty"`
+
+}
+
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=controlplanes,shortName=cp,scope=Namespaced,categories=cluster-api
+// +kubebuilder:storageversion
+// +kubebuilder:subresource:status
+// +kubebuilder:subresource:scale:specpath=.spec.replicas,statuspath=.status.replicas,selectorpath=.status.selector
+
+// ControlPlane is the Schema for the ControlPlane API
+type ControlPlane struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec   ControlPlaneSpec   `json:"spec,omitempty"`
+    Status ControlPlaneStatus `json:"status,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// ControlPlaneList contains a list of ControlPlane
+type ControlPlaneList struct {
+    metav1.TypeMeta `json:",inline"`
+    metav1.ListMeta `json:"metadata,omitempty"`
+    Items           []ControlPlane `json:"items"`
+}
+
+func init() {
+    SchemeBuilder.Register(&ControlPlane{}, &ControlPlaneList{})
+}
+
+```
+
+With the following validations:
+
+- If the referenced InitTemplate does not define external etcd (webhook):
+  - ControlPlane.Spec.Replicas in [1, 3, 5, 7]
+- ControlPlane.Spec.Version != "" (openapi)
+
+And the following defaulting:
+
+- ControlPlaneSpec.Replicas: 1 (openapi or webhook)
+
+#### Modifications required to existing API Types
+
+- Add Cluster.Spec.ControlPlaneRef defined as:
+
+```go
+    // ControlPlaneRef is an optional reference to a provider-specific resource that holds
+    // the details for provisioning the Control Plane for a Cluster
+    // +optional
+    ControlPlaneRef *corev1.ObjectReference `json:"controlPlaneRef,omitempty"`
+```
+
+- Add Cluster.Status.ControlPlaneReady defined as:
+
+```go
+    // ControlPlaneReady defines if the control plane is ready
+    // +optional
+    ControlPlaneReady bool `json:"controlPlaneReady"`
+```
+
+#### Behavioral Changes from v1alpha2
+
+- If Cluster.Spec.ControlPlaneRef is set:
+  - [Status.ControlPlaneInitialized](https://github.com/kubernetes-sigs/cluster-api/issues/1243) is set based on the value of Status.Initialized for the referenced resource
+  - Status.ControlPlaneReady is set based on the value of Status.Ready for the referenced resource, this field is intended to eventually replace Status.ControlPlaneInitialized as a field that will be kept up to date instead of set only once.
+- Current behavior will be preserved if Cluster.Spec.ControlPlaneRef is not set
+
+#### Behaviors
+
+##### Create
+
+- A created ControlPlane should bootstrap a control plane with a given number of replicas
+- If an error occurs, errorStatus and errorReason will be defined on the ControlPlane object
+- Cannot create an even number of stacked control planes or anything other than 1, 3, 5, or 7
+- Can create an arbitrary number of control planes if etcd is external to the control plane
```

Comment:
Is that important in terms of status? One of the additional goals is:

> - To support pre-existing, user-managed external etcd clusters

But it doesn't indicate that support for external etcd systems includes reporting on the health or status of these external resources.

Although one could argue this point **does** link the two:

>  This should include at least the state of etcd and information about which replicas are currently healthy or not.

So if "state of etcd" is desired and "external etcd" is supported, then one could argue whether etcd is external or not should be part of status.

Reply:
I agree that we should do some level of checking even for external etcd, however I'd like to defer that until we get into the implementation rather than fully hash it out now.